Build internal Debian repository

ycointe opened the pull request:
**Component**: Repository, container

<!-- E.g. 'salt', 'containers', 'kubernetes', 'build', 'tests'... -->

**Context**: MetalK8s on Ubuntu

**Summary**: We need to build a Debian repository to dispense Debian packages required for Ubuntu nodes deployment.

**Acceptance criteria**: 
- We need to create a script to execute the `reprepro` command to build a repository inside a container.

**Tests**: (Need to bue done after ISO buit with `./doit.sh`)
- go on the branch `improvement/1617-implement-Debian-repository`
- build Docker image from the Debian Dockerfile (`docker build -t myubuntu:18.04 packages/debian/` )
- Lauch the container with the good volumes mounted:
  `docker run --rm -it -v $(pwd)/packages/debian/entrypoint.sh:/entrypoint.sh -v $(pwd)/packages/debian/distributions:/distributions -v $(pwd)/_build/root/packages/debian/:/metalk8s-deb myubuntu:18.04 bash`
- copie the file `distributions` in the good path (it will be automatically mounted after):
`cp distributions metalk8s-deb/`
- Then lauch the script entrypoint.sh calling `buildrepo` : `./entrypoint.sh buildrepo`

- Once the execution finished, verify in the folder `/repository/debian/dists/bionic/metalk8s-*` that the file `Release` is well here ! (the packages are in the folder `/repository/debian/pool`)
---

<!-- Declare one or more issues to close once this PR gets merged -->

Closes: 

<!-- If you want to refer to an issue while not closing it, use:

See: #ISSUE_NUMBER

-->

ycointe committed:
build/repository: Install reprepro

We need the package `reprepro` install on the container to create a
debian repository.

Refs: #1617

ycointe committed:
build/repository: create a script to build repository

We script the repository build in the entrypoint.

Refs: #1617

bert-e commented:
# Hello ycointe,



My role is to assist you with the merge of this
pull request. Please type `@bert-e help` to get information
on this process, or consult the [user documentation](
https://eve.devsca.com/github/scality/metalk8s/bert-e/doc/user).





*Status report is not available.*

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @slaperche-scality

* @gdemonet

* @carlito-scality

* @sayf-eddine-scality

* @nootal

* @thomasdanan

* @alexandre-allard-scality

* @lucieleonard

* @zempashi

* @monpote

* @jbertran

* @cmonfort

* @nicolast

* @teddyandrieux

slaperche-scality commented on packages/debian/Dockerfile:
```diff
@@ -11,8 +11,8 @@ RUN apt update -y \
         devscripts \
         build-essential \
         lintian \
-        apt-rdepends \
```
Ha good catch

Forgot to remove this one when I rewrite the download in Python.

ycointe replied:
I commit this change in a separate commit ;)

slaperche-scality commented on packages/debian/entrypoint.sh:
```diff
@@ -35,6 +35,19 @@ rpm2deb() {
+    # Mount `distributions` file in  /var/www/debian/conf
+    pushd /var/www/debian
+    reprepro -b . export
+    # We download packages that we want
```
We download?

ycointe replied:
Comment deleted

slaperche-scality commented on packages/debian/entrypoint.sh:
```diff
@@ -35,6 +35,19 @@ rpm2deb() {
+    # A folder `pool` is created whith all .deb insinde
+
+
+## TODO : Handle the GPG key or find a way to do without.
```
Are we OK "with a way without", I'm not sure: I guess we want signed stuff for security purpose.
To be discussed with the team.

ycointe committed:
build/package: remove apt-rdepends

We don't need apt-rdepends in the Dockerfile because the Debian packages
download is handled by `download_packages.py`

Refs: 1617

ycointe committed:
build/repository: Create the file required to create repository

We need a `distributions` file to paste in /repository/debian/conf/ in
the conatiner to create a Debian repository with reprepro.

Refs: #1617

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -35,6 +35,19 @@ rpm2deb() {
+    for dir in /metalk8s-deb/metalk8s-*; do
+        reprepro -C ${dir##*/} -b /var/www/debian includedeb bionic "$dir"/*.deb
+    done
+
```
It's missing the closing curly brace.

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @slaperche-scality

* @gdemonet

* @carlito-scality

* @sayf-eddine-scality

* @nootal

* @thomasdanan

* @alexandre-allard-scality

* @lucieleonard

* @zempashi

* @monpote

* @jbertran

* @cmonfort

* @nicolast

* @teddyandrieux





The following reviewers are expecting changes from the author, or must review again:

* @alexandre-allard-scality

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -35,6 +35,19 @@ rpm2deb() {
     chown "${TARGET_UID}:${TARGET_GID}" *
 }
 
+buildrepo(){
```
details: missing space after the closing parenthesis

alexandre-allard commented on packages/debian/entrypoint.sh:
```diff
@@ -35,11 +35,28 @@ rpm2deb() {
+    for dir in /metalk8s-deb/metalk8s-*; do
+        reprepro -C ${dir##*/} -b /var/www/debian includedeb bionic "$dir"/*.deb
+    done
+
```
extra space

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @slaperche-scality

* @gdemonet

* @carlito-scality

* @sayf-eddine-scality

* @nootal

* @thomasdanan

* @alexandre-allard-scality

* @lucieleonard

* @zempashi

* @monpote

* @jbertran

* @cmonfort

* @nicolast

* @teddyandrieux

gdemonet commented on packages/debian/entrypoint.sh:
```diff
@@ -36,10 +36,27 @@ rpm2deb() {
+    mkdir -p /var/www/debian/{conf,incoming}
+    # We assume that all downloaded packages are mounted in /metalk8s-deb/
+    cp /metalk8s-deb/distributions /var/www/debian/conf
+    # Mount `distributions` file in  /var/www/debian/conf
```
It's not a mount, you're copying it. IMO, the comment can be removed.

ycointe replied:
The file `distributions` is created before the repository build, it should be mounted in the container then, paste in the folder `/var/www/debian/conf`.
This comment is just a reminder for the next PR, integrate this script in the buildchain (it will be deleted after)

gdemonet commented on packages/debian/entrypoint.sh:
```diff
@@ -36,10 +36,27 @@ rpm2deb() {
 
+buildrepo() {
+    set -x
+    mkdir -p /var/www/debian/{conf,incoming}
```
Why use `/var/www/debian`? I'm confused, is it supposed to be served by some webserver?

gdemonet replied:
Also, if you want to set a basedir for `reprepro`, instead of duplicating the `-b` argument for each invocation, you could just set the `REPREPRO_BASE_DIR` environment variable.

ycointe replied:
we donc need the `/var/www` for now, I will remove it

gdemonet commented on packages/debian/distributions:
```diff
@@ -0,0 +1,7 @@
+Codename: bionic
+Architectures: amd64
+Components: main metalk8s-bionic metalk8s-bionic-backports metalk8s-bionic-updates metalk8s-kubernetes-xenial metalk8s-salt_ubuntu1804 metalk8s-scality
+Description: Build repository for Debian
```
Not sure this description is accurate. It should describe the repository, so let's go for something like:

```
Description: MetalK8s repository for bionic
```

ycointe replied:
Indeed, it's from my tests, I will change it.

gdemonet commented on packages/debian/entrypoint.sh:
```diff
@@ -36,10 +36,27 @@ rpm2deb() {
+    cp /metalk8s-deb/distributions /var/www/debian/conf
+    # Mount `distributions` file in  /var/www/debian/conf
+    reprepro -b /var/www/debian export
+    for dir in /metalk8s-deb/metalk8s-*; do
```
So, IIUC (this loop would benefit from a comment explaining what it does, for readers unaware of Debian repositories building practices), we're adding all packages for `metalk8s-*` components.

Don't we also need to do the same for the `main` component? At least that's what I would expect from the `distributions` file.

ycointe replied:
In fact there is no main component, I will remove it from the `distributions` file.

gdemonet commented on packages/debian/distributions:
```diff
@@ -0,0 +1,7 @@
+Suite: stable
+Codename: bionic
+Architectures: amd64
+Components: main metalk8s-bionic metalk8s-bionic-backports metalk8s-bionic-updates metalk8s-kubernetes-xenial metalk8s-salt_ubuntu1804 metalk8s-scality
```
Why not `metalk8s-main`?

ycointe replied:
In fact I have to delete main... there is no need of this component

gdemonet reviewed (dismissed):
Bunch of questions, would like to have some comments in the code to get more info on this process.

Also, how can we make sure this script works, if it's not yet used in the buildchain? Would help to catch errors, and help reviewers to test what it's doing (also, we can't rely on CI since it won't hit this code).

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @slaperche-scality

* @gdemonet

* @carlito-scality

* @sayf-eddine-scality

* @nootal

* @thomasdanan

* @alexandre-allard-scality

* @lucieleonard

* @zempashi

* @monpote

* @jbertran

* @cmonfort

* @nicolast

* @teddyandrieux





The following reviewers are expecting changes from the author, or must review again:

* @gdemonet

bert-e commented:
# Waiting for approval



The following approvals are needed before I can proceed with the merge:


* the author


* one peer





Peer approvals *must* include at least 1 approval from the following list:

* @slaperche-scality

* @gdemonet

* @carlito-scality

* @sayf-eddine-scality

* @nootal

* @thomasdanan

* @alexandre-allard-scality

* @lucieleonard

* @zempashi

* @monpote

* @jbertran

* @cmonfort

* @nicolast

* @teddyandrieux

ycointe commented:
/approve

bert-e commented:
# Build failed



The build for [commit](https://github.com/scality/metalk8s/commit/712cbe82fa2eec3fb6616b89eb97bf75f8d104a8) did not succeed in branch improvement/1617-implement-Debian-repository.




*The following options are set:* **approve**

bert-e commented:
# In the queue



The changeset has received all authorizations and has been added to the
relevant queue(s). The queue(s) will be merged in the target development
branch(es) as soon as builds have passed.

The changeset will be merged in:


* :heavy_check_mark: `development/2.4`



The following branches will **NOT** be impacted:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



There is no action required on your side. You will be notified here once
the changeset has been merged. In the unlikely event that the changeset
fails permanently on the queue, a member of the admin team will
contact you to help resolve the matter.

*IMPORTANT*

Please do not attempt to modify this pull request.

* Any commit you add on the source branch will trigger a new cycle after the
  current queue is merged.
* Any commit you add on one of the integration branches will be **lost**.

If you need this pull request to be removed from the queue, please contact a
member of the admin team now.




*The following options are set:* **approve**

bert-e commented:
# 


I have successfully merged the changeset of this pull request
into targetted development branches:


* :heavy_check_mark: `development/2.4`



The following branches have **NOT** changed:

* `development/1.0`
* `development/1.1`
* `development/1.2`
* `development/1.3`
* `development/2.0`
* `development/2.1`
* `development/2.2`
* `development/2.3`



Please check the status of the associated issue None.

Goodbye ycointe.
